### PR TITLE
Unify offset configuration in DDIM and PNDM schedulers

### LINKED ISSUE #465: Discrepancy in scheduler `offset` usage
**What API design would you like to have changed or added to the library? Why?**

@jonatanklosko brought up an issue with some confusing code (why are `alphas` indexed differently across `PNDMScheduler` and `DDIMScheduler`), and some inconsistency in the offset usage.

Two questions:
1. Why does only one shift by offset?
2. Can `offset` be changed from an arbitrary `int` to a `bool` setting that shifts by 1 (the integer shift is more confusing if the only use case is stable diffusion inference with `offset = 1`)

Discussion began in #444, see [here](https://github.com/huggingface/diffusers/issues/444#issuecomment-1242811446)

https://github.com/huggingface/diffusers/blob/8eaaa546d89f836b716e92348786d878f883ee86/src/diffusers/schedulers/scheduling_pndm.py#L338-L339

https://github.com/huggingface/diffusers/blob/8eaaa546d89f836b716e92348786d878f883ee86/src/diffusers/schedulers/scheduling_ddim.py#L204-L205

I can tell you that the PNDMScheduler and DDIMScheduler have indexed the alpha's differently for a while. Before introducing offset, PNDMScheduler had

```
alpha_prod_t = self.alphas_cumprod[timestep + 1]
alpha_prod_t_prev = self.alphas_cumprod[timestep_prev + 1]
```

And DDIMScheduler had

```
alpha_prod_t = self.alphas_cumprod[timestep]
alpha_prod_t_prev = self.alphas_cumprod[timestep_prev]
```

At minimum, we can figure this out and document better!

### PR DESCRIPTION
**Pull request by jonatanklosko**
Closes #465.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 1]
```diff
@@ -217,12 +217,7 @@ def __call__(
         latents = latents.to(self.device)
 
         # set timesteps
-        accepts_offset = "offset" in set(inspect.signature(self.scheduler.set_timesteps).parameters.keys())
-        extra_set_kwargs = {}
-        if accepts_offset:
-            extra_set_kwargs["offset"] = 1
-
-        self.scheduler.set_timesteps(num_inference_steps, **extra_set_kwargs)
+        self.scheduler.set_timesteps(num_inference_steps)
```
Comment (jonatanklosko) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:220]: Should we try to override the configuration here, or are you going to update the configuration in the relevant hf repositories?

File: src/diffusers/schedulers/scheduling_pndm.py [hunk 2]
```diff
@@ -88,8 +96,10 @@ def __init__(
         beta_end: float = 0.02,
         beta_schedule: str = "linear",
         trained_betas: Optional[np.ndarray] = None,
-        tensor_format: str = "pt",
         skip_prk_steps: bool = False,
+        set_alpha_to_one: bool = True,
```
Comment (jonatanklosko) [inline src/diffusers/schedulers/scheduling_pndm.py:100]: Ah, I guess we want `False` for backward compatibility, even though it's not in line with DDIM?

File: src/diffusers/schedulers/scheduling_pndm.py [hunk 3]
```diff
@@ -335,8 +355,8 @@ def _get_prev_sample(self, sample, timestep, timestep_prev, model_output):
         # sample -> x_t
         # model_output -> e_θ(x_t, t)
         # prev_sample -> x_(t−δ)
-        alpha_prod_t = self.alphas_cumprod[timestep + 1 - self._offset]
-        alpha_prod_t_prev = self.alphas_cumprod[timestep_prev + 1 - self._offset]
```
Comment (jonatanklosko) [inline src/diffusers/schedulers/scheduling_pndm.py:339]: @patrickvonplaten this change breaks this test:

https://github.com/huggingface/diffusers/blob/f4781a0b27ffb3ea61ecd25b0b87305e0960304e/tests/test_scheduler.py#L633-L655

However, I think the test must be wrong, because there we use the default `offset=0`, which means that we are using `self.alphas_cumprod[1]` for the timestep `0`, which doesn't make sense. I assume the value in the test is computed using the current implementation, rather than a reference value from elsewhere?

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 4]
```diff
@@ -135,15 +138,25 @@ def _get_variance(self, timestep, prev_timestep):
...
                 the number of diffusion steps used when generating samples with a pre-trained model.
-            offset (`int`): TODO
         """
+
+        offset = self.config.steps_offset
+
+        if "offset" in kwargs:
+            warnings.warn(
+                "`offset` is deprecated as an input argument to `set_timesteps` and will be removed in v0.4.0."
+                " Please pass `steps_offset` to `__init__` instead."
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_ddim.py:155]: nice !

File: src/diffusers/schedulers/scheduling_pndm.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_pndm.py:100]: Yes we need `False` for backward compability indeed :-)

File: src/diffusers/schedulers/scheduling_pndm.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_pndm.py:339]: Hmm, I think the test breaks because `set_alpha_to_one` defaults to `True` no?

**Review by patrickvonplaten — COMMENTED** [review]
Super cool PR! Very nice that you've deprecated the `offset` function argument - that's very nice! 

Think we're very close to merging this one :-) Think there are just two things to do:

- Make sure the change is fully backwards compatible
- Add a couple of tests to DDIM and PNDM that show how `set_alpha_to_one` and `offset` should be passed to `__init__` and changes the outputs respectively (happy to help here if you want :-) )

File: tests/test_scheduler.py [hunk 5]
```diff
@@ -398,26 +426,31 @@ def test_variance(self):
         assert torch.sum(torch.abs(scheduler._get_variance(999, 998) - 0.02)) < 1e-5
 
     def test_full_loop_no_noise(self):
```
Comment (jonatanklosko) [inline tests/test_scheduler.py:428]: @patri in https://github.com/huggingface/diffusers/commit/394243ce98c79e273732f8e09b2e5c50ee9a9bd7 you change PNDM step to accept the timestep value rather than iteration number, but I think the tests were kept as is. I changed the full loop to use actual timesteps.

File: tests/test_scheduler.py [hunk 6]
```diff
@@ -631,28 +699,30 @@ def test_inference_plms_no_past_residuals(self):
             scheduler.step_plms(self.dummy_sample, 1, self.dummy_sample).prev_sample
 
     def test_full_loop_no_noise(self):
```
Comment (jonatanklosko) [inline tests/test_scheduler.py:701]: @patri in https://github.com/huggingface/diffusers/commit/394243ce98c79e273732f8e09b2e5c50ee9a9bd7 you change PNDM step to accept the timestep value rather than iteration number, but I think the tests were kept as is. I changed the full loop to use actual timesteps.

File: tests/test_scheduler.py [hunk 7]
```diff
@@ -606,8 +659,23 @@ def test_timesteps(self):
...
+        assert torch.equal(
+            scheduler.timesteps,
+            torch.tensor(
+                [901, 851, 851, 801, 801, 751, 751, 701, 701, 651, 651, 601, 601, 501, 401, 301, 201, 101, 1]
+            ),
+        )
+
     def test_betas(self):
-        for beta_start, beta_end in zip([0.0001, 0.001, 0.01], [0.002, 0.02, 0.2]):
+        for beta_start, beta_end in zip([0.0001, 0.001], [0.002, 0.02]):
```
Comment (jonatanklosko) [inline tests/test_scheduler.py:678]: I removed this betas combination as it leads to infinite results now with the update to `step_prk`.

**Comment by jonatanklosko** [comment]
@patrickvonplaten I added some test, let me know what you think. I hardcoded the values that are currently returned, or do usually obtain them in a different way?

> Make sure the change is fully backwards compatible

Anything I should test in particular?

File: src/diffusers/schedulers/scheduling_pndm.py (hunk 3 above)
Comment (jonatanklosko) [reply src/diffusers/schedulers/scheduling_pndm.py:341]: I kept the change and adjusted the test value.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 8]
```diff
@@ -190,6 +183,7 @@ def __call__(
         init_latents = torch.cat([init_latents] * batch_size)
 
         # get the original timestep using init_timestep
+        offset = self.scheduler.config.get("steps_offset", 0)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:186]: Nice!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 9]
```diff
@@ -53,6 +54,21 @@ def __init__(
     ):
         super().__init__()
         scheduler = scheduler.set_format("pt")
+
+        if hasattr(scheduler.config, "steps_offset") and scheduler.config.steps_offset != 1:
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:58]: @jonatanklosko - Currently the slow tests are broken because we haven't (and can't yet) update the configs online. So for now I think we need to do some deprecation warning here - ok for you?

**Review by patrickvonplaten — APPROVED** [review]
Amazing PR! 

Sorry for fiddling into your code here @jonatanklosko - only added some code to make all the stability pipelines backwards compatible. 
Once this code is in tho `0.4.0` release, we can update all the configs online on the Hub and nudge people to upgrade their `diffusers` version. 

@anton-l @patil-suraj @pcuenca - please have a look here as well. We need to make sure that we remember to update all the configs on the Hub once this PR is in the next release. Also once we update the configs old diffusers versions should throw a warning because of an unused parameter.

**Comment by patrickvonplaten** [comment]
@jonatanklosko - let me if the PR is good for you as it's now!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 9 above)
Comment (jonatanklosko) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:58]: Perfect!

Our of curiosity, is there an issue with updating the configs now? I assume an additional property in the config doesn't hurt even if ignored by previous versions?

**Comment by jonatanklosko** [comment]
@patrickvonplaten fantastic, thanks for fixing the pipelines! Looks good to me :)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 9 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:58]: Most users are on the `0.3.0` version therefore will not make use of this code **but** will nevertheless use the newest config file. Given the high usage of stable diffusion: https://huggingface.co/CompVis/stable-diffusion-v1-4 (>500k downloads per month), I think it would be better to not burden users with an extra config parameter that should throw a warning in `0.3.0` .

So, it would indeed not break anything, but an unused config parameter should throw a warning (haven't checked though whether our code actually works correctly here), which will probably scare users.

Does that make sense?

Comment (jonatanklosko) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:58]: Ah totally, I didn't realise the extra option would cause a warning :)
